### ISSUE TITLE
Class and Interface return types incompatible

### ISSUE THREAD

**Issue by dani-h** [Microsoft/TypeScript#1633]
I'm new to Typescript so bear with me

```javascript
interface I {
  func: () => void
  func2: () => void
}

class A implements I {

  //Compiles fine, even though the return should be detected by the compiler?
  func = () => {
    return "Hello";
  }

  //Bug.ts(6,7): error TS2421: Class 'A' incorrectly implements interface 'I':
  //Types of property 'func2' are incompatible:
  //Type '(s: string) => void' is not assignable to type '() => void'.
  func2 = (s:string) => {

  }
}
```

func2 makes sense, it shows errors when compiling as the implementation shouldn't have any parameters.
func however, compiles fine even though it returns a string and the return type should be void. Shouldn't the compiler detect errors here?

**Comment by DanielRosenwasser** [comment]
This is because we define override compatibility with respect to assignability.

From [8.2.3 of the spec](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#823-inheritance-and-overriding)

> A property member in a derived class is said to override a property member in a base class when the derived class property member has the same name and kind (instance or static) as the base class property member. The type of an overriding property member **must be assignable** (section [3.10.4](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#3.10.4)) to the type of the overridden  property member, or otherwise a compile-time error occurs.

And from [3.10.4](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#3.10.4):

> *S* is ***assignable to*** a type *T*, and *T* is ***assignable from*** *S*, if one of the following is true:
* ...
* *S* is an object type, a type parameter, or the Number, Boolean, or String primitive type, *T* is an object type, and for each member *M* in *T*, one of the following is true:
  * ...
  * *M* is a non-specialized call or construct signature and *S* has an apparent call or construct signature *N* where, when *M* and *N* are instantiated using type Any as the type argument for all type parameters declared by *M* and *N* (if any),
    * ...
    * the result type of *M* is Void, or the result type of *N* is assignable to that of *M*.

The general intuitionistic approach to assignability is that you can always give "more" when "less" is expected (and you can always take "more" when you expect "less"). So, this assignability rule is safe because you can't actually do anything useful with `undefined` or` null` (apart from differentiate between them), so it is permissible to return "more" (i.e. a `string`).

**Comment by mhegazy** [comment]
following up on @DanielRosenwasser explanation, if you define your parameter (s: string) as optional (s?: string) in the class that will make it match the interface definition:

```TypeScript
func2 = (s?:string) => {
}
```

**Comment by dani-h** [comment]
@mhegazy So if the interface declares that the function takes no parameters, and the implementation takes optional parameters it will compile fine? That seems a little backwards, shouldn't the class always comply with the interface? I.e if I wanted a function with optional parameters I could specify that in the interface.

**Comment by RyanCavanaugh** [comment]
> So if the interface declares that the function takes no parameters, and the implementation takes optional parameters it will compile fine?

It's hard to imagine why you *wouldn't* want that to be the case. Consider something like this:
```ts
interface ToStringable {
  toString(): string;
}

class MyNumber implements ToStringable {
  toString(radix?: number) {
    if(radix !== undefined) { ... }
  }
}
```

**Comment by dani-h** [comment]
Seeing as it's part of the spec I'll close this.
